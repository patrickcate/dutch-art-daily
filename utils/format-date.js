import isToday from 'date-fns/is_today'
import isYesterday from 'date-fns/is_yesterday'
import subDays from 'date-fns/sub_days'
import addDays from 'date-fns/add_days'

/**
 * Returns the date ordinal string base on the numeric day of the month.
 * @param {Number} day
 */
export function getDateOrdinal(day) {
  if (day === 1 || day === 21 || day === 31) return 'st'
  if (day === 2 || day === 22) return 'nd'
  if (day === 3 || day === 23) return 'rd'
  return 'th'
}

/**
 * Takes a ISO 8601 formatted date string and formats it in the correct format
 * for display as the page date title.
 * @param {String} ISODate
 */
export function formatDate(id, part) {
  // Parse the ISO 8601 string into an actual Date object.
  const date = new Date(getISODate(id))

  // Setup date formatting options.
  // @see https://mzl.la/1TepFRT
  const options = {}
  if (part === 'month') {
    options.month = 'long'
  } else if (part === 'day') {
    options.day = 'numeric'
  } else {
    options.month = 'long'
    options.day = 'numeric'
  }

  // Format the date with the optional prefix.
  return new Intl.DateTimeFormat('en-US', options).format(date)
}

export function getISODate(id) {
  // Convert a artwork date id to an ISO date string.
  return getDateObjFromId(id).toISOString()
}

export function getDateObjFromId(id) {
  // Break the data parameters into an array so we can access the month
  // and day separately.
  const date = id.split('-')
  const year = new Date().getFullYear()
  const dateObject = new Date(year, date[0] - 1, date[1])

  return dateObject
}

/**
 * Takes a JavaScript date object and converts it to a string that
 * matches a date id.
 * @param { Date } dateObj
 */
export function dateId(dateObj) {
  // Get the month for the previous day. We need to add 1 since months are
  // 0 based. Also, we pad months < 10 with a 0.
  const month = `${dateObj.getMonth() + 1}`.padStart(2, '0')

  // Get the day of the month for the previous day.
  // Also, we pad months < 10 with a 0.
  const day = `${dateObj.getDate()}`.padStart(2, '0')

  // Return the month and day as a URL string.
  return `${month}-${day}`
}

export function dateShort(dateObj) {
  return `${dateObj.getMonth() + 1}/${dateObj.getDate()}`
}

export function yesterdayDateId(dateObj) {
  return dateId(subDays(dateObj, 1))
}

export function tomorrowDateId(dateObj) {
  return dateId(addDays(dateObj, 1))
}

export function dateArray(dateId) {
  return dateId.split('-')
}

export function dateObject(dateId) {
  const parts = dateArray(dateId)

  return {
    month: parts[0],
    day: parts[1],
  }
}

export function getDateLabel(id) {
  const isoDate = getISODate(id)

  // Check if the date matches today's date.
  if (isToday(isoDate)) {
    return 'Today'
  }
  // Check if the date matches yesterdays's date.
  else if (isYesterday(isoDate)) {
    return 'Yesterday'
  }

  return ''
}

export function generateDateList(id) {
  const today = getDateObjFromId(id)
  const dateList = [id]

  for (let i = 1; i < 7; i++) {
    dateList.push(dateId(subDays(today, i)))
  }

  return dateList.reverse()
}
