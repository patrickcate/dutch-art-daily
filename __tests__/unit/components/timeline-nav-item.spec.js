import { dateId } from '@utils/format-date'
import TimelineNavItem from '@components/timeline-nav-item.vue'
import { getters as defaultGetters } from '~/store/index.js'

describe('TimelineNavItem Component', () => {
  let wrapper
  let options

  beforeEach(() => {
    options = {
      propsData: {
        id: '01-01',
      },
      computed: {
        breakpoint() {
          return {
            md: true,
          }
        },
      },
    }

    wrapper = createWrapper(TimelineNavItem, options)
  })

  it('should be a Vue instance', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders with date label', () => {
    const todaysDateId = dateId(new Date())

    wrapper = createWrapper(TimelineNavItem, {
      propsData: {
        id: todaysDateId,
      },
      computed: {
        breakpoint() {
          return {
            md: true,
          }
        },
      },
    })

    expect(wrapper).toMatchSnapshot()
  })

  it('goes to clicked slide', async () => {
    const slideToMock = jest.fn()

    wrapper = createWrapper(
      TimelineNavItem,
      {
        ...options,
        propsData: {
          id: '12-30',
        },
      },
      {
        getters: defaultGetters,
      }
    )

    wrapper.vm.$root.swipers.carousel = {
      slideTo(index) {
        slideToMock(index)
      },
    }

    wrapper.find('button').trigger('click')
    await wrapper.vm.$nextTick()

    expect(slideToMock).toHaveBeenCalledWith(4)
  })

  it('stays on current slide when clicked', async () => {
    const slideToMock = jest.fn()

    wrapper = createWrapper(
      TimelineNavItem,
      {
        ...options,
        propsData: {
          id: '01-01',
        },
      },
      {
        getters: defaultGetters,
      }
    )

    wrapper.vm.$root.swipers.carousel = {
      slideTo(index) {
        slideToMock(index)
      },
    }

    wrapper.find('button').trigger('click')
    await wrapper.vm.$nextTick()

    expect(slideToMock).not.toHaveBeenCalled()
  })
})
