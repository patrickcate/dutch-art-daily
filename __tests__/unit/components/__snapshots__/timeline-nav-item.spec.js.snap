// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineNavItem Component renders correctly 1`] = `
<li class="timeline-nav__list-item"><button type="button" class="line-segment">
    <div class="line-segment__accent"></div>
    <div class="line-segment__point"></div>
    <div class="line-segment__label"><span class="u-sr-only">Go to </span> <span>01/01</span> <span class="u-sr-only"> slide</span></div>
  </button></li>
`;

exports[`TimelineNavItem Component renders with date label 1`] = `
<li class="timeline-nav__list-item"><button type="button" class="line-segment">
    <div class="line-segment__accent"></div>
    <div class="line-segment__point"></div>
    <div class="line-segment__label"><span class="u-sr-only">Go to </span> <span class="line-segment__prefix">Today</span> <span class="u-sr-only"> slide</span></div>
  </button></li>
`;
